sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,366 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// A <see cref="PipelinePolicy"/> used by a <see cref="ClientPipeline"/> to
+/// log request and response information.
+/// </summary>
+public class ClientLoggingPolicy : PipelinePolicy
+{
+    private const double RequestTooLongSeconds = 3.0; // sec
+
+    private readonly bool _logContent;
+    private readonly int _maxLength;
+    private readonly string _clientAssembly = "System-ClientModel";
+    private readonly LoggingHandler _handler;
+    private readonly ILogger _logger;
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="ClientLoggingPolicy"/> class.
+    /// </summary>
+    /// <param name="clientAssembly">The assembly name to include with each entry.</param>
+    /// <param name="options">The user-provided logging options object.</param>
+    protected ClientLoggingPolicy(string clientAssembly, LoggingOptions? options = default) : this(options)
+    {
+        _clientAssembly = clientAssembly;
+    }
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="ClientLoggingPolicy"/> class.
+    /// </summary>
+    /// <param name="options">The user-provided logging options object.</param>
+    public ClientLoggingPolicy(LoggingOptions? options = default)
+    {
+        LoggingOptions loggingOptions = options ?? new LoggingOptions();
+        _logContent = loggingOptions.IsLoggingContentEnabled;
+        _maxLength = loggingOptions.LoggedContentSizeLimit;
+        PipelineMessageSanitizer sanitizer = new(loggingOptions.AllowedQueryParameters.ToArray(), loggingOptions.AllowedHeaderNames.ToArray());
+        _logger = loggingOptions.LoggerFactory.CreateLogger("System.ClientModel");
+        _handler = new LoggingHandler(_logger, sanitizer);
+    }
+
+    /// <inheritdoc/>
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+
+    /// <inheritdoc/>
+    public override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex) =>
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        if (!(ClientModelEventSource.Log.IsEnabled() || _logger.IsEnabled(LogLevel.Warning))) // The highest LogLevel we log with is LogLevel.Warning
```

Comment:
A downside with this approach in general is that if someone passes in an ILogger and then adds a listener, it may be confusing why they don't get events written to the event source.